diff --git a/Cargo.toml b/Cargo.toml
index 1963289..45305ea 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -34,7 +34,7 @@ alloy-rpc-types-eth = "0.3"
 alloy-rpc-types-trace = "0.3"
 alloy-sol-types = "0.8"
 alloy-primitives = "0.8"
-revm = { version = "14", default-features = false, features = ["std"] }
+revm = { path = "../revm/crates/revm", default-features = false, features = ["std"] }
 
 anstyle = "1.0"
 colorchoice = "1.0"
